### ISSUE TITLE
-- noqa ignored in multiline comment

### ISSUE THREAD

**Issue by daviewales** [sqlfluff/sqlfluff#4891]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have a very long URL inside a multiline comment.
I don't want to trigger a rule failure, so I tried putting `-- noqa: LT05` at the end of the line.
But because we are already inside a multiline comment, the `noqa` declaration was ignored.

### Expected Behaviour

There should be a way to specify `noqa` for comment lines. (See also #4881 which would be a better solution in this particular case.)

### Observed Behaviour

`sqlfluff lint` outputs the following failure even when `-- noqa: LT05` is appended to the long comment line:

    L:   2 | P:   1 | LT05 | Line is too long (106 > 88).

### How to reproduce

Run `sqlfluff lint` on the following SQL file:

``` sql
/*
* https://example.com/1111111111111111111111111111111111111111111111111111111111111111111111 -- noqa: LT05
*/
```

### Dialect

tsql

### Version

sqlfluff, version 2.1.0
Python 3.10.10

### Configuration

``` cfg
[sqlfluff]
dialect = tsql
max_line_length = 88

# ST06 reorders columns, which seems presumptuous
exclude_rules = ST06

[sqlfluff:indentation]
allow_implicit_indents = True

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True
additional_allowed_characters = ['-/()?,']
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by daviewales** [comment]
I've tracked it down to this line:
https://github.com/sqlfluff/sqlfluff/blob/7e4843427b8e34ccdd0909b55f16842159027f80/src/sqlfluff/core/linter/linter.py#L451
If I just remove that check, it happily follows `-- noqa` directives at the end of lines inside multiline comments.
I'm not sure if this has unintended consequences however.

From what I can tell, each line of a multiline comment is passed one at a time to `parse_noqa()`, which allows the function to process it the same as an inline comment.

Happy to make a PR with this change if there are no objections.

**Comment by Robert-Kosmala-FLYR** [comment]
hi @daviewales,

I would be also interested in allowing noqa directives in multiline comments.
in our case we often copy SQL text from the logs where it is kept as a single line.
single line SQL comments are breaking such SQL text and we do not allow ourselves to use them.
I would then welcome an option to use directives in below form:
```
/* noqa: disable=AL02 */
SELECT col_a a FROM foo
/* noqa: enable=AL02 */
```

would it be possible for the PR you're thinking of to cover this scenario as well?

**Comment by daviewales** [comment]
I did a quick check, and the small change I suggested above doesn't appear to enable your use-case.

It _does_ work if you put the `noqa` on its own line inside the multiline comment.
But doesn't work if you have a single line 'multiline' comment:

i.e. With the change suggested above, this does what you want:

``` sql
SELECT col_a a FROM foo
/*
noqa: disable=AL02
*/
SELECT col_a a FROM foo
/*
noqa: enable=AL02
*/
SELECT col_a a FROM foo
```

But this doesn't:

``` sql
SELECT col_a a FROM foo
/* noqa: disable=AL02 */
SELECT col_a a FROM foo
/* noqa: enable=AL02 */
SELECT col_a a FROM foo
```

I'd need to dig a bit deeper into the code to understand why it doesn't enable the second form, but I'll wait to hear if the PR would be accepted before working on it.

**Comment by Robert-Kosmala-FLYR** [comment]
thanks for checking this. yes, I think for it to be coherent, both syntaxes need to be covered.
it's already 2 months since you've initially posted the idea. is there a way we could request a feedback here or it's in some queue and we should just wait?

**Comment by daviewales** [comment]
I'm just waiting. It's an open source project, so the maintainers are likely doing this in their own time.

**Comment by alanmcruickshank** [comment]
I'm working through a few things first - but then might be able to pick this one up. I'm also happy to provide some pointers if anyone else is up for contributing a fix?

**Comment by daviewales** [comment]
@alanmcruickshank I'm happy to have a crack at it if you give me some pointers.

Would you accept a PR to make the following test cases work?

1. ``` sql
    /*
    * https://example.com/1111111111111111111111111111111111111111111111111111111111111111111111 -- noqa: LT05
    */
    ```

2. ``` sql
    SELECT col_a a FROM foo
    /*
    noqa: disable=AL02
    */
    SELECT col_a a FROM foo
    /*
    noqa: enable=AL02
    */
    SELECT col_a a FROM foo
    ```

3. ```
    SELECT col_a a FROM foo
    /* noqa: disable=AL02 */
    SELECT col_a a FROM foo
    /* noqa: enable=AL02 */
    SELECT col_a a FROM foo
    ```

Do you have some pointers how this should work?
Does it make sense that `-- noqa: LT05` at the end of a line in a multiline comment should be followed to allow specifying `noqa` for comment lines?
Are you happy with the other syntax suggestions above?

**Comment by alanmcruickshank** [comment]
@daviewales - I'm comfortable with the syntax suggestions you've made 👍 .

As for pointers:
- The current parsing of `noqa` lines is done [here ](https://github.com/sqlfluff/sqlfluff/blob/b90620caeba4749b1c2fe8999cb04dc4fef86576/src/sqlfluff/core/linter/linter.py#L266C15-L266C15). You'll see that the method returns a `NoQaDirective` object on success containing the information about the command. At the moment the comment has to _start_ with `noqa` after taking off any `--`.
- The other issue is that the method above is effectively called from [here](https://github.com/sqlfluff/sqlfluff/blob/b90620caeba4749b1c2fe8999cb04dc4fef86576/src/sqlfluff/core/linter/linter.py#L442). You'll note that it's only called for segments of type `inline_comment`. You'll need to open that up to also cover segments of type `block_comment` to resolve this issue.

My expectation is that if you do both of those things, and handle any weird edge cases that arise as a result - you'll be good 👍

**Comment by alanmcruickshank** [comment]
@daviewales - heads up that I'm working on something in a similar area in parallel (#2337), which shouldn't mess up anything you're doing, but I have just put together a PR which _moves_ some of the code (methods all still intact though). It's not merged yet, but you might want to take a look at #5010 if you've already started working on this as we'll probably merge it in the next few days.

I hope it doesn't cause too much disruption to your work. 🤞

**Comment by daviewales** [comment]
Thanks, very considerate of you to mention it. I haven't started yet, so I might just wait until you merge it.

**Comment by alanmcruickshank** [comment]
> Thanks, very considerate of you to mention it. I haven't started yet, so I might just wait until you merge it.

Merged 👍 . All yours now.

**Comment by alanmcruickshank** [comment]
To remain consistent with other noqa comments on the end of long _single line_ comments, I think we should still require the --. Does that seem sensible?

Btw if you haven't already seen, I was in some of this code over the last few days putting together this PR https://github.com/sqlfluff/sqlfluff/pull/5029. I don't think it changes the plan here but it might have moved some of the code.

**Comment by daviewales** [comment]
Thanks @alanmcruickshank, that makes sense.

For consistency, should we also require `--` when the `noqa` is the only thing in the comment?
For example:

**without `--`**:

``` sql
 SELECT col_a a FROM foo
 /* noqa: disable=AL02 */
 SELECT col_a a FROM foo
 /* noqa: enable=AL02 */
 SELECT col_a a FROM foo
```

**with `--`**:

``` sql
 SELECT col_a a FROM foo
 /* -- noqa: disable=AL02 */
 SELECT col_a a FROM foo
 /* -- noqa: enable=AL02 */
 SELECT col_a a FROM foo
```

**Comment by alanmcruickshank** [comment]
When it starts a comment I think we can be more lenient. I think either should work 👍
